Add scrobbling support via SLS

f1u77y opened the pull request:
Fixes #326 and also adds scrobbling support to other supported services
(eg. libre.fm).

The reasoning behind not implementing the interaction with the service(s) in this app are described [here](https://github.com/marverenic/Jockey/issues/6#issuecomment-358376949).

Here's the link to [SLS API description](https://github.com/tgwizard/sls/blob/master/Developer's%20API.md).

marverenic commented on app/src/main/java/com/marverenic/music/player/MusicPlayer.java:
```diff
@@ -679,6 +680,7 @@ public void pause() {
         if (isPlaying()) {
             mMediaPlayer.pause();
             updateNowPlaying();
+            new SlsInteraction(mContext).sendPause(getNowPlaying());
```
I'm a bit concerned about putting this logic here because, although simple, it adds yet another responsibility to `MusicPlayer`. The main goal of the class is to provide a high-level implementation of a music player – so things like shuffling, repeat modes, and integration with Android audio frameworks like audio focus and media session.

Scrobbling to Last.fm isn't _really_ critical to the behavior of a music player – it's at a higher level than this class, and if you're not careful about this sort of thing then all of a sudden one file ends up doing everything related to music playback in a very fragile and unmaintainable way (Take a look at the [single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle) if you're interested in more info about this).

The way I'd go about this is to make a new interface (maybe called `MusicPlayerExtension`?) with various callbacks for player events like `onPlay`, `onPause`, etc. Then you can give `MusicPlayer` an `addExtension` and `removeExtension` method that adds/removes a callback. With this in place, all of the behavior to send Last.fm events can live in one file. (Also, if you do make a `MusicPlayerExtension` interface, it's probably better to make it an abstract class with empty method bodies so that extensions don't have to implement every single callback if they don't need them).

f1u77y replied:
> I'm a bit concerned

I was concerned about this too but there was nothing (except the `OnPlaybackChangeListener` which didn't suit for this). I agree that making such an interface is better idea.

marverenic commented on app/src/main/java/com/marverenic/music/utils/SlsInteraction.java:
```diff
@@ -0,0 +1,53 @@
+
+import com.marverenic.music.model.Song;
+
+public class SlsInteraction {
```
Nit on the name here: instances of this classes aren't single interactions – they're delegates that send an interaction. Maybe `SlsMessenger` would make this more clear?

It might also be good to have a quick Javadoc for the class that links to https://github.com/tgwizard/sls, just because SLS isn't a commonly used acronym.

f1u77y replied:
>  Maybe `SlsMessenger` would make this more clear?

I'd assume that `SlsMessenger` could be used more than once and `SlsInteraction` to  be used not more than once. This class also represents a single interaction. So, I think `SlsInteraction` is more appropriate name for this.

marverenic commented on app/src/main/java/com/marverenic/music/utils/SlsInteraction.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    private Intent withSong(Song song) {
+        return base()
```
That's a good point – I hadn't thought about async issues because was in the mindset of immutable objects and builders. This should be fine as-is then, since there isn't really any performance hit for allocating Intent objects here.

f1u77y commented:
I've fixed all the things you said, could you review again?

About background execution limits: as far as I can understand, it's fine if the reciever app registers the explicit action. SLS [does](https://github.com/tgwizard/sls/blob/master/app/src/main/AndroidManifest.xml#L108). So, that's not a problem, is it?

marverenic commented on app/src/main/java/com/marverenic/music/utils/SlsMessenger.java:
```diff
@@ -0,0 +1,87 @@
+
+/**
+ * A helper class for interacting with Simple Last.fm Scrobbler. The API description could be found
+ * <a href=https://github.com/tgwizard/sls/blob/master/Developer's%20API.md>here</a>.
```
The rest of the project generally has the convention of inserting the URL directly instead of using an HTML tags, since HTML Javadocs aren't actually generated for this project.

marverenic commented on app/src/main/java/com/marverenic/music/utils/SlsMessenger.java:
```diff
@@ -0,0 +1,87 @@
+                .putExtra(KEY_ALBUM, song.getAlbumName())
+                .putExtra(KEY_ARTIST, song.getArtistName())
+                .putExtra(KEY_TRACK, song.getSongName())
+                .putExtra(KEY_DURATION, (int)TimeUnit.MILLISECONDS.toSeconds(song.getSongDuration()))
```
Super nit: can we add a space after the cast here?

marverenic commented on app/src/main/java/com/marverenic/music/utils/SlsMessenger.java:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * Should be called when a song is just completed. It's not necessary to send this just after
+     * sending Start for a new song but it should be sent at the end of the track if the next track
+     * is the same for proper scrobbling
```
I'm having a bit of trouble following this Javadoc. Can you explain the second sentence?

f1u77y replied:
When you play A then B, you don't need to send Complete and scrobbling will be done fine but if you play C on repeat, you should send Complete each time song is completed for this song to be scrobbled more than once. Anyway, don't think that's a useful note.

marverenic commented on app/src/main/java/com/marverenic/music/player/MusicPlayerExtension.java:
```diff
@@ -0,0 +1,10 @@
+import com.marverenic.music.model.Song;
+
+public abstract class MusicPlayerExtension {
+    void onSongStart(Song song) {}
```
Can we rename this to `onSongStarted` to be consistent with the rest of the tenses here?

marverenic commented on app/src/main/java/com/marverenic/music/player/ScrobblerExtension.java:
```diff
@@ -0,0 +1,34 @@
+package com.marverenic.music.player;
```
I'm wondering if this should go in a separate package along with `SlsMessenger`, or if there should be an `com.marverenic.music.player.extensions` package. Thoughts?

marverenic commented:
(This also needs a merge master)

f1u77y commented:
> It's a bit out of the scope of this PR, but I think it'd also be good to give users an option to turn this on or off.

SLS already gives such an option, so I don't think it's useful.

> (This also needs a merge master)

I'll squash and rebase when you approve.

marverenic commented:
I'm actually thinking from more of a security and performance perspective: an application could fairly easily impersonate SLS. Plus if the user wanted to turn this off, then there would be a (very) small battery and performance win since we're not spinning up processes. That's not something I think should block this PR from being merged, though.

I also usually prefer to wait until a PR is completely ready to merge before giving approval, which requires being up-to-date with master. It looks like there's a merge conflict (not sure if that's new since I last reviewed), but once it's resolved this should be ready to merge.

f1u77y commented:
> I'm actually thinking from more of a security and performance perspective: an application could fairly easily impersonate SLS. Plus if the user wanted to turn this off, then there would be a (very) small battery and performance win since we're not spinning up processes.
What does "impersonate" mean here? A malicious app that eg. sends this data to some unwanted destination? Don't think it's a big security issue because data on last.fm/libre.fm is publicly availible anyway. But you're right that it's just better not to send data from the app when user doesn't wants it to do this.

About Oreo: as far as I can understand, if an app statically registers explicit reciever, only the same app can send explicit intents to this reciever. Anyway, if it's not in this way, I don't get the point of explicit intents.

I've fixed all of you said before, squashed and rebased though.

I've

marverenic commented:
Yeah, there's nothing stopping an app from declaring a broadcast receiver for the `com.adam.aslfms.notify.playstatechanged` action. It's definitely not the worst thing to leak someone's music history, but some people will mind.

With explicit intents, I don't know that they're applicable to broadcasts – typically the implicit vs. explicit terminology comes when you have a specific action and you resolve that you're going to start it with a specific service/activity. Compared to an implicit intent which Android will show the "Select an app to perform this action" disambiguation dialog if there's a default app. Neither one of those particularly applies here; we're just blindly broadcasting it to all apps. This is something SLS will have to consider, though, so for now there's not too much we can do besides wait.

All that being said, this looks good to me. Thanks for the contribution, and feel free to let me know if you want a name drop in the release notes or anything.
